apis/python/src/tiledbsoma/experiment_query/query.py

```diff
@@ -0,0 +1,503 @@
+        """
+        Clean up and close all resources. This must be called or the thread pool
+        will not be released.
+        """
+        if self.__threadpool is not None:
+            self.__threadpool.shutdown()
+            self.__threadpool = None
+
+    def __enter__(self) -> "ExperimentAxisQuery":
+        return self
+
+    def __exit__(self, *excinfo: Any) -> None:
+        self.close()
+
+    def obs(
+        self, *, column_names: Optional[Sequence[str]] = None
+    ) -> somacore.ReadIter[pa.Table]:
+        """Return obs as an Arrow table iterator."""
+        query = self._query["obs"]
+        return self.experiment.obs.read(
+            ids=query.coords,
+            value_filter=query.value_filter,
+            column_names=column_names,
+        )
+
+    def var(
+        self, *, column_names: Optional[Sequence[str]] = None
+    ) -> somacore.ReadIter[pa.Table]:
+        """Return obs as an Arrow table iterator."""
+        query = self._query["var"]
+        return self.experiment.ms[self.measurement_name].var.read(
+            ids=query.coords,
+            value_filter=query.value_filter,
+            column_names=column_names,
+        )
+
+    def obs_joinids(self) -> pa.Array:
+        """Return obs soma_joinids as an Arrow array."""
+        return self._read_axis_joinids("obs", self.experiment.obs)
+
+    def var_joinids(self) -> pa.Array:
+        """Return var soma_joinids as an Arrow array."""
+        return self._read_axis_joinids(
+            "var", self.experiment.ms[self.measurement_name].var
+        )
+
+    @property
+    def n_obs(self) -> int:
+        """Return number of obs axis query results"""
+        return len(self.obs_joinids())
+
+    @property
+    def n_vars(self) -> int:
+        """Return number of var axis query results"""
+        return len(self.var_joinids())
+
+    def X(self, layer_name: str) -> SparseNDArrayRead:
+        """
+        Return an X layer as SparseNDArrayRead.
+
+        Parameters
+        ----------
+        layer_name : str
+            The X layer name to return.
+
+        Examples
+        --------
+        >>> with experiment.query(
+        ...     "RNA",
+        ...     obs_query=AxisQuery(value_filter='tissue == "lung"')
+        ... ) as query:
+        ...     X_result = query.X("raw").tables().concat()
+        >>> X_result
+        pyarrow.Table
+        soma_dim_0: int64
+        soma_dim_1: int64
+        soma_data: float
+        ----
+        soma_dim_0: [[1538112,1538112,1538112,1538112,1538112,...,1538308,1538308,1538308,1538308,1538308], ...]
+        soma_dim_1: [[5,19,26,34,37,...,10577,10603,10616,10617,10655], ...]
+        soma_data: [[2,2,1,1,1,...,1,1,1,2,2], ...]
+        """
+        if not (
+            layer_name and layer_name in self.experiment.ms[self.measurement_name].X
+        ):
+            raise ValueError("Must specify X layer name")
+
+        X = self.experiment.ms[self.measurement_name].X[layer_name]
```

Comment:
I know that we love `X` as a member name and an array name, but I’m still not a fan of using capitals for variables.